component/discovery/transformer/transformer.go

```diff
@@ -0,0 +1,265 @@
+		if rc.SourceLabels != nil ||
+			rc.TargetLabel != DefaultRelabelConfig.TargetLabel ||
+			rc.Modulus != DefaultRelabelConfig.Modulus ||
+			rc.Separator != DefaultRelabelConfig.Separator ||
+			rc.Replacement != DefaultRelabelConfig.Replacement {
+			return errors.Errorf("%s action requires only 'regex', and no other fields", rc.Action)
+		}
+	}
+
+	return nil
+}
+
+// Exports holds values which are exported by the discovery.transformer component.
+type Exports struct {
+	OutputTargets []Target `hcl:"output_targets,attr"`
+}
+
+// Component implements the discovery.transformer component.
+type Component struct {
+	opts component.Options
+
+	mut  sync.Mutex
+	args Arguments
+
+	healthMut sync.RWMutex
+	health    component.Health
+}
+
+var (
+	_ component.Component       = (*Component)(nil)
+	_ component.HealthComponent = (*Component)(nil)
+)
+
+// New creates a new discovery.transformer component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{opts: o}
+
+	// Call to Update() to set the output once at the start
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	<-ctx.Done()
+	return nil
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.args = newArgs
+
+	targets := make([]Target, 0, len(newArgs.Targets))
+	failures := make([]error, 0, len(newArgs.Targets))
+	relabelConfigs, err := hclToPromRelabelConfigs(newArgs.RelabelConfigs)
+	if err != nil {
+		return err
+	}
+
+	for _, t := range newArgs.Targets {
+		var cfg = &config.ScrapeConfig{
+			RelabelConfigs: relabelConfigs,
+			Params:         newArgs.QueryParams,
+			JobName:        newArgs.ScrapeLabels.JobName,
+			MetricsPath:    newArgs.ScrapeLabels.MetricsPath,
+			Scheme:         newArgs.ScrapeLabels.Scheme,
+			ScrapeInterval: model.Duration(newArgs.ScrapeLabels.Interval),
+			ScrapeTimeout:  model.Duration(newArgs.ScrapeLabels.Timeout),
+		}
+		lbls, _, err := scrape.PopulateLabels(hclMapToPromLabels(t), cfg)
+		if err != nil {
+			failures = append(failures, errors.Wrapf(err, "failed to correctly mutate target with labelset: %s", t))
+		}
+		if lbls != nil {
+			targets = append(targets, promLabelsToHCL(lbls))
+		}
+	}
+
+	c.opts.OnStateChange(Exports{
+		OutputTargets: targets,
+	})
+
+	c.setHealth(component.Health{
+		Health:     component.HealthTypeHealthy,
+		Message:    "set output targets",
+		UpdateTime: time.Now(),
+	})
+
+	return nil
+}
+
+// CurrentHealth implements component.HealthComponent.
+func (c *Component) CurrentHealth() component.Health {
+	c.healthMut.RLock()
+	defer c.healthMut.RUnlock()
+	return c.health
+}
+
+func (c *Component) setHealth(h component.Health) {
+	c.healthMut.Lock()
+	defer c.healthMut.Unlock()
+	c.health = h
+}
```

Comment:
I think you can remove this; the controller will update the health of the component when evaluations finish, and component-specific health can be reserved for runtime specific problems (which this component wouldn't have).

Reply:
Oh, I see! In any case, we should leave `CurrentHealth()` intact so that the component implements the HealthComponent interface, right?